File: src/commands/pr.ts
import { $ } from "bun";
import { findGwtRoot, checkGwtSetup, getWorktrees } from "../core/repo";
import { resolveWorktree, selectWorktree } from "./cd";

export async function pr(action?: string, name?: string): Promise<void> {
  const check = checkGwtSetup();
  if (!check.ok) {
    throw new Error(`Error: ${check.error}`);
  }

  const root = findGwtRoot()!;
  process.chdir(root);

  const ghCheck = await $`command -v gh`.quiet().nothrow();
  if (ghCheck.exitCode !== 0) {
    throw new Error("Error: 'gh' CLI not found. Install it from https://cli.github.com");
  }

  const worktrees = await getWorktrees();
  if (worktrees.length === 0) {
    throw new Error("No worktrees found");
  }

  const wt = name ? resolveWorktree(worktrees, name) : await selectWorktree(worktrees);

  if (!wt.branch) {
    throw new Error("Error: Worktree is in detached HEAD state");
  }

  if (action === "create") {
    console.log(`Creating PR for '${wt.branch}'...`);
    const result = await $`gh pr create --web --head ${wt.branch}`.quiet().nothrow();
    if (result.exitCode !== 0) {
      const stderr = result.stderr.toString().trim();
      throw new Error(`Error: Failed to create PR\n${stderr}`);
    }
  } else {
    const result = await $`gh pr view ${wt.branch} --web`.quiet().nothrow();
    if (result.exitCode !== 0) {
      const stderr = result.stderr.toString().trim();
      if (stderr.includes("no pull requests found")) {
        throw new Error(`No PR found for branch '${wt.branch}'. Use 'gwt pr create' to create one.`);
      }
      throw new Error(`Error: Failed to view PR\n${stderr}`);
    }
  }
}

File: src/pr.rs
use std::ffi::OsString;

use crate::AppResult;
use crate::arg_to_str;
use crate::cd::{resolve_worktree, select_worktree};
use crate::repo::{command_exists, ensure_gwt_setup, get_worktrees, run_command};

pub fn run(args: &[OsString]) -> AppResult<()> {
    if !command_exists("gh") {
        return Err("Error: 'gh' CLI not found. Install it from https://cli.github.com".to_string());
    }
    let (action, name) = parse_options(args)?;
    let root = ensure_gwt_setup()?;
    let worktrees = get_worktrees(&root)?;
    if worktrees.is_empty() {
        return Err("No worktrees found".to_string());
    }
    let worktree = if let Some(name) = name.as_deref() {
        resolve_worktree(&worktrees, name)?.clone()
    } else {
        select_worktree(&worktrees)?
    };
    let branch = worktree
        .branch
        .ok_or_else(|| "Error: Worktree is in detached HEAD state".to_string())?;

    if action.as_deref() == Some("create") {
        println!("Creating PR for '{}'...", branch);
        let result = run_command("gh", &["pr", "create", "--web", "--head", &branch], None)?;
        if result.exit_code != 0 {
            return Err(format!("Error: Failed to create PR\n{}", result.stderr.trim()));
        }
    } else {
        let result = run_command("gh", &["pr", "view", &branch, "--web"], None)?;
        if result.exit_code != 0 {
            let stderr = result.stderr.trim();
            if stderr.contains("no pull requests found") {
                return Err(format!("No PR found for branch '{}'. Use 'gwt pr create' to create one.", branch));
            }
            return Err(format!("Error: Failed to view PR\n{}", stderr));
        }
    }

    Ok(())
}

fn parse_options(args: &[OsString]) -> AppResult<(Option<String>, Option<String>)> {
    let mut action = None;
    let mut name = None;
    let mut index = 0;

    while index < args.len() {
        match arg_to_str(&args[index])? {
            "-w" => {
                index += 1;
                let value = args.get(index).ok_or_else(|| "Error: -w requires a worktree name".to_string())?;
                name = Some(arg_to_str(value)?.to_string());
            }
            value if value.starts_with('-') => return Err(format!("Error: unknown pr option '{value}'")),
            value => {
                if action.is_none() && value == "create" {
                    action = Some(value.to_string());
                } else if name.is_none() {
                    name = Some(value.to_string());
                } else {
                    return Err("Error: too many pr arguments".to_string());
                }
            }
        }
        index += 1;
    }

    Ok((action, name))
}
